File: app/src/main/java/net/ballmerlabs/uscatterbrain/db/entities/ApiIdentity.java
package net.ballmerlabs.uscatterbrain.db.entities;

import androidx.annotation.Nullable;

import com.google.protobuf.ByteString;
import com.goterl.lazycode.lazysodium.interfaces.Sign;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import net.ballmerlabs.uscatterbrain.API.Identity;
import net.ballmerlabs.uscatterbrain.network.IdentityPacket;
import net.ballmerlabs.uscatterbrain.network.LibsodiumInterface;

import java.util.ConcurrentModificationException;
import java.util.HashMap;
import java.util.Map;
import java.util.SortedSet;
import java.util.TreeSet;

public class ApiIdentity extends Identity {
    private byte[] privatekey;

    protected ApiIdentity(Builder builder) {
        super(
                builder.mPubKeymap,
                builder.mPubKeymap.get(IdentityPacket.PROTOBUF_PRIVKEY_KEY),
                builder.name,
                builder.sig
        );

        this.privatekey = builder.privkey;
    }

    @Nullable
    public byte[] getPrivateKey() {
        return privatekey;
    }

    public static class KeyPair {
        public final byte[] secretkey;
        public final byte[] publickey;
        private KeyPair(byte[] sec, byte[] pub) {
            this.publickey = pub;
            this.secretkey = sec;
        }
    }

    public static KeyPair newPrivateKey() {
        byte[] sec = new byte[Sign.SECRETKEYBYTES];
        byte[] pub = new byte[Sign.PUBLICKEYBYTES];

        LibsodiumInterface.getSodium().crypto_sign_keypair(pub, sec);

        return new KeyPair(sec, pub);
    }

    public static Builder newBuilder() {
        return new Builder();
    }

    public static class Builder {
        private byte[] sig;
        private Map<String, byte[]> mPubKeymap = new HashMap<>();
        private String name;
        private byte[] pubkey;
        private byte[] privkey;
        private KeyPair signPair;
        private Builder() {
            super();
        }

        private ByteString sumBytes() {
            ByteString result = ByteString.EMPTY;
            result = result.concat(ByteString.copyFromUtf8(name));
            SortedSet<String> sortedKeys= new TreeSet<>(mPubKeymap.keySet());
            for (String key : sortedKeys) {
                result = result.concat(ByteString.copyFromUtf8(key));
                final byte[] k = mPubKeymap.get(key);
                if (k == null) {
                    throw new ConcurrentModificationException();
                }
                result = result.concat(ByteString.copyFrom(k));
            }
            return result;
        }

        /**
         * Sign ed 25519 boolean.
         *
         * @param secretkey the secretkey
         * @return the boolean
         */
        private synchronized boolean signEd25519(byte[] secretkey) {
            if (secretkey.length != Sign.SECRETKEYBYTES)
                return false;

            ByteString messagebytes = sumBytes();

            byte[] signature = new byte[Sign.ED25519_BYTES];
            Pointer p = new PointerByReference(Pointer.NULL).getPointer();
            if (LibsodiumInterface.getSodium().crypto_sign_detached(signature,
                    p, messagebytes.toByteArray(), messagebytes.size(), secretkey) == 0) {
                this.sig = signature;
                return true;
            } else {
                return false;
            }
        }

        public Builder setName(String name) {
            this.name = name;
            return this;
        }

        public Builder sign(KeyPair keyPair) {
            this.signPair = keyPair;
            return this;
        }

        public Builder setSig(byte[] sig) {
            this.sig = sig;
            return this;
        }

        public Builder addKeys(Map<String, byte[]> keys) {
            this.mPubKeymap.putAll(keys);
            return this;
        }

        public ApiIdentity build() {
            if (name == null) {
                throw new IllegalArgumentException("name should be non-null");
            }

            if (sig == null && signPair == null) {
                throw new IllegalArgumentException("sig should be set");
            }

            if (sig != null && signPair != null) {
                throw new IllegalArgumentException("cannot sign and set sig simultaneously");
            }

            if (signPair != null) {
                signEd25519(signPair.secretkey);
                mPubKeymap.put(IdentityPacket.PROTOBUF_PRIVKEY_KEY, signPair.publickey);
                this.pubkey = signPair.publickey;
            } else {
                if (!mPubKeymap.containsKey(IdentityPacket.PROTOBUF_PRIVKEY_KEY)) {
                    throw new IllegalArgumentException("key map does not contain scatterbrain pubkey");
                }

                pubkey = mPubKeymap.get(IdentityPacket.PROTOBUF_PRIVKEY_KEY);
            }
            return new ApiIdentity(this);
        }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/db/entities/ApiIdentity.kt
package net.ballmerlabs.uscatterbrain.db.entities

import com.google.protobuf.ByteString
import com.goterl.lazysodium.interfaces.GenericHash
import com.goterl.lazysodium.interfaces.Sign
import com.sun.jna.Pointer
import com.sun.jna.ptr.PointerByReference
import kotlinx.collections.immutable.toImmutableMap
import net.ballmerlabs.scatterbrainsdk.Identity
import net.ballmerlabs.scatterproto.*
import net.ballmerlabs.uscatterbrain.network.LibsodiumInterface
import net.ballmerlabs.uscatterbrain.util.hashAsUUID
import java.util.SortedSet
import java.util.TreeSet
import java.util.UUID

/**
 * ApiIdentity is a mutable handle to an identity that allows more privileged
 * read/write access including the ability to sign and modify the private key
 * This is used for working with identities internally
 */
open class ApiIdentity protected constructor(val builder: Builder) {
    private val privatekey: ByteArray? = builder.privkey
    val privateKey: ByteArray?
        get() = privatekey


    val identity: Identity
    get() = Identity(
        builder.mPubKeymap.toImmutableMap(),
        builder.mPubKeymap[PROTOBUF_PRIVKEY_KEY]!!,
        builder.name!!,
        builder.sig!!,
        builder.fingerprint!!,
        builder.hasPrivateKey
    )


    /**
     * keypairs are effectively a tuple of an ed25519 public and private key
     */
    class KeyPair(val secretkey: ByteArray, val publickey: ByteArray)

    class Builder {
        var sig: ByteArray? = null
        val mPubKeymap: MutableMap<String, ByteArray> = HashMap()
        var name: String? = null
        private var pubkey: ByteArray? = null
        var privkey: ByteArray? = null
        private var signPair: KeyPair? = null
        var fingerprint: UUID? = null
        var hasPrivateKey = false
        private fun sumBytes(): ByteString {
            var result = ByteString.EMPTY
            result = result.concat(ByteString.copyFromUtf8(name))
            val sortedKeys: SortedSet<String> = TreeSet(mPubKeymap.keys)
            for (key in sortedKeys) {
                result = result.concat(ByteString.copyFromUtf8(key))
                val k = mPubKeymap[key] ?: throw ConcurrentModificationException()
                result = result.concat(ByteString.copyFrom(k))
            }
            return result
        }

        private fun getPubkeyFingerprint(): UUID {
            val fingeprint = ByteArray(GenericHash.BYTES)
            LibsodiumInterface.sodium.crypto_generichash(
                    fingeprint,
                    fingeprint.size,
                    pubkey,
                    pubkey!!.size.toLong(),
                    null,
                    0
            )
            return hashAsUUID(fingeprint)
        }

        /**
         * Sign ed 25519 boolean.
         *
         * @param secretkey the secretkey
         * @return the boolean
         */
        private fun signEd25519(secretkey: ByteArray): Boolean {
            if (secretkey.size != Sign.SECRETKEYBYTES) return false
            val messagebytes = sumBytes()
            val signature = ByteArray(Sign.ED25519_BYTES)
            val p = PointerByReference(Pointer.NULL).pointer
            return if (LibsodiumInterface.sodium.crypto_sign_detached(signature,
                            p, messagebytes.toByteArray(), messagebytes.size().toLong(), secretkey) == 0) {
                sig = signature
                true
            } else {
                false
            }
        }

        fun setName(name: String): Builder {
            this.name = name
            return this
        }

        fun sign(keyPair: KeyPair): Builder {
            signPair = keyPair
            return this
        }

        fun setSig(sig: ByteArray): Builder {
            this.sig = sig
            return this
        }

        fun addKeys(keys: Map<String, ByteArray>): Builder {
            mPubKeymap.putAll(keys)
            return this
        }
        
        
        fun setHasPrivateKey(hasPrivKey: Boolean): Builder {
            hasPrivateKey = hasPrivKey
            return this
        }

        fun build(): ApiIdentity {
            requireNotNull(name) { "name should be non-null" }
            require(!(sig == null && signPair == null)) { "sig should be set" }
            require(!(sig != null && signPair != null)) { "cannot sign and set sig simultaneously" }
            if (signPair != null) {
                require(signPair?.publickey != null) { "public key must be non-null" }
                require(signPair?.secretkey != null) { "secret key must be non-null" }
                mPubKeymap[PROTOBUF_PRIVKEY_KEY] = signPair!!.publickey
                pubkey = signPair!!.publickey
                privkey = signPair!!.secretkey
                signEd25519(signPair!!.secretkey)
                hasPrivateKey = true
            } else {
                pubkey = mPubKeymap[PROTOBUF_PRIVKEY_KEY]
            }
            fingerprint = getPubkeyFingerprint()
            return ApiIdentity(this)
        }
    }

    companion object {
        fun newPrivateKey(): KeyPair {
            val sec = ByteArray(Sign.SECRETKEYBYTES)
            val pub = ByteArray(Sign.PUBLICKEYBYTES)
            LibsodiumInterface.sodium.crypto_sign_keypair(pub, sec)
            return KeyPair(sec, pub)
        }

        fun newBuilder(): Builder {
            return Builder()
        }
    }

}